File: Seori/Programmers/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    answer = 0
    rest_delivery, rest_pickup = 0, 0
    # [1] 가장 먼 집부터 순차적으로 방문한다.
    #     집 now에서 해결해야할 택배 수는 주어진 값(deliveries, pickups) - 더 먼 집에서 남은 값(rest_delivery, rest_pickup)이다.
    for now in range(n-1, -1, -1):
        round = 0
        next_delivery, next_pickup = deliveries[now] - rest_delivery, pickups[now] - rest_pickup
        
        # [2-1] 해결할 택배가 있는 경우는 배달/수거 대소비교를 통해 더 큰 값만큼 왕복 횟수 round를 정한다.
        if next_delivery >= next_pickup and next_delivery > 0:
            round = ((next_delivery - 1) // cap + 1)
        elif next_pickup > next_delivery and next_pickup > 0:
            round = ((next_pickup - 1) // cap + 1)
        
        # [2-2] 해결할 택배가 없는 경우는 round가 0이 되어 거리 계산을 생략한다.
        answer += 2 * round * (now + 1)
        
        # [2-3] 이후 round만큼 왕복하면서 남은 배달/수거값은 각각 rest_delivery, rest_pickup 변수에 저장한다.
        rest_delivery = cap * round - next_delivery
        rest_pickup = cap * round - next_pickup
            
    return answer

File: Lucy/2025/2-W01/택배_배달과_수거하기.js
function solution(cap, n, deliveries, pickups) {
  let answer = 0;

  const deliveryStack = [];
  const pickupStack = [];

  for (let i = 0; i < n; i++) {
    const currentDelivery = deliveries[i];
    const currentPickup = pickups[i];

    if (currentDelivery > 0) {
      deliveryStack.push({ position: i, cap: currentDelivery });
    }
    if (currentPickup > 0) {
      pickupStack.push({ position: i, cap: currentPickup });
    }
  }

  while (deliveryStack.length > 0 || pickupStack.length > 0) {
    const peakDelivery = deliveryStack[deliveryStack.length - 1];
    const peakPickup = pickupStack[pickupStack.length - 1];

    if (peakDelivery && peakPickup) {
      answer += (Math.max(peakDelivery.position, peakPickup.position) + 1) * 2;
    } else if (peakDelivery) {
      answer += (peakDelivery.position + 1) * 2;
    } else if (peakPickup) {
      answer += (peakPickup.position + 1) * 2;
    }

    let truck = 0;
    while (deliveryStack.length > 0 && truck < cap) {
      if (truck + deliveryStack[deliveryStack.length - 1].cap > cap) {
        deliveryStack[deliveryStack.length - 1].cap -= cap - truck;
        truck = cap;

        break;
      } else {
        truck += deliveryStack.pop().cap;
      }
    }

    truck = 0;
    while (pickupStack.length > 0 && truck < cap) {
      if (truck + pickupStack[pickupStack.length - 1].cap > cap) {
        pickupStack[pickupStack.length - 1].cap -= cap - truck;
        truck = cap;

        break;
      } else {
        truck += pickupStack.pop().cap;
      }
    }
  }

  return answer;
}

solution(4, 5, [1, 0, 3, 1, 2], [0, 3, 0, 4, 0]); // 16
solution(2, 7, [1, 0, 2, 0, 1, 0, 2], [0, 2, 0, 1, 0, 2, 0]); // 30
solution(3, 5, [2, 3, 4, 1, 2], [0, 0, 0, 0, 0]); // 26

// function solution(cap, n, deliveries, pickups) {
//   let deliveriesPointer = deliveries.findLastIndex((delivery) => delivery > 0);
//   let pickupsPointer = pickups.findLastIndex((pickup) => pickup > 0);

//   let res = 0;
//   while (deliveriesPointer >= 0 && pickupsPointer >= 0) {
//     if (deliveries.reduce((acc, curr) => acc + curr, 0) === 0 && pickups.reduce((acc, curr) => acc + curr, 0) === 0) break;
//     res += (deliveriesPointer + 1) * 2;

//     let truck = 0;

//     for (let i = deliveriesPointer; i >= 0; i--) {
//       const currentDelivery = deliveries[i];

//       if (truck + currentDelivery > cap) {
//         deliveriesPointer = i;

//         break;
//       }

//       if (currentDelivery <= cap) {
//         deliveries[i] = 0;
//         truck += currentDelivery;
//       } else {
//         deliveries[i] -= cap;
//         truck += cap;

//         deliveriesPointer = i;
//         break;
//       }
//     }

//     truck = 0;

//     for (let i = pickupsPointer; i >= 0; i--) {
//       const currentPickup = pickups[i];

//       if (truck + currentPickup > cap) {
//         pickupsPointer = i;
//         truck = 0;

//         break;
//       }

//       if (currentPickup <= cap) {
//         pickups[i] = 0;
//         truck += currentPickup;
//       } else {
//         pickups[i] -= cap;
//         truck += cap;

//         pickupsPointer = i;
//         truck = 0;
//         break;
//       }
//     }
//   }

//   console.log(res);
//   return res;
// }

// solution(4, 5, [1, 0, 3, 1, 2], [0, 3, 0, 4, 0]); // 16
// solution(2, 7, [1, 0, 2, 0, 1, 0, 2], [0, 2, 0, 1, 0, 2, 0]); // 30
